### ISSUE TITLE
[Bug?] Modality sampling granularity in _sample_modalities differs from Algorithm 1 in the paper

### ISSUE THREAD

**Issue by mmosc** [Tigxy/SiBraR---Single-Branch-Recommender#7]
Hi Christian!

Could it be that we had a mismatch in the code compared to what we described in our paper?

Algorithm 1 in the paper specifies that modalities are sampled once per interaction (one draw per user–item pair in the batch), and the same sampled set $\mathcal{M}_{ij}$ is reused for the positive item and all negative items of that interaction:

```
M_ij ← randomSampling(M^item)          # sampled once
e_j  ← (1/n_mod) Σ_{m ∈ M_ij} g(m_j)  # positive item, uses M_ij
for i_k ∈ N_ij:
    e_k ← (1/n_mod) Σ_{m ∈ M_ij} g(m_k)  # each negative, reuses M_ij
```

However, the code seems to implement something different:

 In `SingleBranchNetEntity._sample_modalities`, items are passed with shape `(batch_size, 1+n_neg)` (positive and negatives stacked), and `row_wise_sample` is called with `size=indices.shape`:

```
# sgd_alg.py
n_rows = math.prod(size)  # = batch_size * (1 + n_neg)
result = np.array([choice_fn(a, k, replace=replace) for i in range(n_rows)])
```
This seems to produces `batch_size * (1+n_neg)` independent draws (one per item) so the positive item and each negative of the same interaction receive different modality pairs.

Therefore, the InfoNCE contrastive loss is applied between modalities that might differ between positive item $i$ and negative items $j$'s.

What do you think? Should we fix that?

Cheers,
Marta

**Comment by Tigxy** [comment]
Hi Marta,
Thanks for the great catch! The modality sampling is indeed different than what we describe in our paper (and also what I thought is implemented). While I could imagine that this makes the training more stable, improves the performance of SiBraR, and perhaps even slightly mitigates the embedding space collapse we observed, for sake of reproducability of our original work, I prefer to keep the code as it for now until we know how this actually affects training.

Nonetheless, here's the updated `SingleBranchNetEntity._sample_modalities` that should now behave as described. If you ever have the chance, could you please try it out? Unfortunately, I am currently lacking the time & resources to do so myself. Once done, I'll close this issue and check in the code in the main branch.

```py
def _sample_modalities(self, indices, reg_type):
        if self.training and self.do_sampling:
            batch_size, *repetitions = indices.shape
            
            common_sampling_params = {
                "a": list(self.train_modalities),
                "size": batch_size,
                "rng": self._rng,
            }

            match self.entity_config.embedding_regularization_type:
                case EmbeddingRegularizationType.NoRegularization:
                    # sample just a single modality per indices
                    mods = row_wise_sample(**common_sampling_params, k=1)

                case EmbeddingRegularizationType.PairwiseSingle:
                    # sample two available modalities per indices
                    mods = row_wise_sample(**common_sampling_params, k=2)

                case EmbeddingRegularizationType.CentralModality:
                    # sample 1 other modality in addition to the central modality
                    mods = row_wise_sample(
                        **common_sampling_params,
                        k=2,
                        # central_item=self.entity_config.central_modality,
                        central_item=self.centr_mod,
                    )
                case _:
                    raise ValueError(
                        f'Embedding regularization "{self.entity_config.embedding_regularization_type}"'
                        f"is not yet supported."
                    )

            if repetitions:
                mods = np.tile(mods[..., *[None]*len(repetitions), :], (*repetitions, 1))
            return mods
            
        else:
            # select all modalities during evaluation (a weighting of them, e.g., their average, should be taken
            mods = np.empty(
                shape=indices.shape + (len(self.eval_modalities),), dtype=object
            )
            mods[None:] = list(self.eval_modalities)
            return mods
```

Thanks again for letting me know!

Best, 
Chris
